contracts/peer-to-peer/oracles/custom/MysoOracle.sol

```diff
@@ -0,0 +1,197 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.19;
+
+import {ChainlinkBase} from "../chainlink/ChainlinkBase.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {Ownable2Step} from "@openzeppelin/contracts/access/Ownable2Step.sol";
+import {IWSTETH} from "../../interfaces/oracles/IWSTETH.sol";
+import {IMETH} from "../../interfaces/oracles/IMETH.sol";
+import {Math} from "@openzeppelin/contracts/utils/math/Math.sol";
+import {IERC20Metadata} from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+
+/**
+ * @dev supports oracles which are compatible with v2v3 or v3 interfaces
+ */
+contract MysoOracle is ChainlinkBase, Ownable2Step {
+    struct MysoPrice {
+        uint112 prePrice;
+        uint112 postPrice;
+        uint32 switchTime;
+    }
+
+    // solhint-disable var-name-mixedcase
+    address internal constant MYSO = 0x00000000000000000000000000000000DeaDBeef; // TODO: put in real myso address
+    address internal constant WETH = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2; // weth
+    address internal constant WSTETH =
+        0x7f39C581F595B53c5cb19bD0b3f8dA6c935E2Ca0; //wsteth
+    address internal constant METH = 0xd5F7838F5C461fefF7FE49ea5ebaF7728bB0ADfa; //meth
+    address internal constant RPL = 0xD33526068D116cE69F19A9ee46F0bd304F21A51f; //rpl
+    address internal constant METH_STAKING_CONTRACT =
+        0xe3cBd06D7dadB3F4e6557bAb7EdD924CD1489E8f; //meth staking contract
+    uint256 internal constant MYSO_IOO_BASE_CURRENCY_UNIT = 1e18; // 18 decimals for ETH based oracles
+    address internal constant ETH_USD_CHAINLINK =
+        0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419; //eth usd chainlink
+    address internal constant STETH_ETH_CHAINLINK =
+        0x86392dC19c0b719886221c78AB11eb8Cf5c52812; //steth eth chainlink
+    address internal constant RPL_USD_CHAINLINK =
+        0x4E155eD98aFE9034b7A5962f6C84c86d869daA9d; //rpl usd chainlink
+
+    uint256 internal constant MYSO_PRICE_TIME_LOCK = 1 hours;
+
+    MysoPrice public mysoPrice;
+
+    event MysoPriceUpdated(
+        uint112 prePrice,
+        uint112 postPrice,
+        uint32 switchTime
+    );
+
+    error NoMyso();
+
+    /**
+     * @dev constructor for MysoOracle
+     * @param _tokenAddrs array of token addresses
+     * @param _oracleAddrs array of oracle addresses
+     * @param _mysoUsdPrice initial price of myso in usd (use 8 decimals like chainlink) (eg. 0.50 USD = 0.5 * 1e8)
+     */
+    constructor(
```

Comment:
maybe we could add the owner as constructor arg such that we can assign it to e.g. multisig/insti custody without having to claim ownership via 2step approval process.